src/Generator/Generators/CSharp/CSharpSources.cs

```diff
@@ -3326,5 +3334,95 @@ public int GetHashCode(Parameter obj)
                 return obj.Type.Desugar().GetHashCode();
             }
         }
+
+        class LibrarySymbolInfo
+        {
+            private static readonly Regex identifierCleanerRegex = new Regex(@"[^\w]", RegexOptions.Compiled);
```

Comment:
You don't need mangled names for identifiers since they're limited in scope (the containing class). For the case of two variables with the same name within a library you already have a counter (or you could prepend their containing class if any). So this regex and all of the related logic is useless.

Reply:
I don't follow. The regex is there to strip characters that can potentially generate invalid identifiers and doing that with such a simple regex will likely generate collisions, so there's code to handle that. If there's useless logic can you be more specific where that is.

Reply:
The invalid characters come from basing identifiers on mangled symbols as I've explained.

Reply:
Got it but I saw the problem from another angle and I'm too tired to change that today. If I prepend the containing class, there could be a collision as well. Then I prepend the namespace or use CSharpTypePrinter.GetName or GetQualifiedName, then I'll have to remove invalid identifiers again. At the end of the day seems like the same work.

Reply:
It's fine, we can continue tomorrow. This will also give me some time to think about the obstacle you've presented.

Reply:
@josetr what do you think about this: you use the name of the variable if it's free and _class_variable_ if it's a static member. In theory this could still lead to collisions but in practice hardly.
Even if you decide it's simpler to use `GetQualifiedName`, you'd only have to replace `::` which is still better than mangling and regexes.

Reply:
In practice isn't good enough tho. I don't know what bothers you most: the ugly names or the extra 4 lines to avoid collisions that will only trigger 1% of the time. The regex will always be executed but it is very trivial and it's compiled as well.

The idea was to create a compile time cache for the symbols, similar to what happens when you compile an executable and in assembly you can many places dereferencing pointers that have not yet been resolved until the dlls are loaded.

You could also imagine an scenario where you create an ILRewriter and replace calls to DllImport methods with `calli`, which seems to be faster. You can use this cache to easily do that.

On a similar subject I also was wondering what's faster: using a custom `ICustomMarshaller` or just marshalling the string yourself and then calling `DllImport` methods with their native types. 

You could also merge it and improve it any way you want it later with another pull request.

Reply:
I don't mind the idea, in fact, I tend to agree with you (and @tritao ) that it would be in general faster. I just avoid making even the generated code difficult to read unless absolutely necessary. So I think we can reach a quick solution: you could keep the regex but apply it to qualified instead of mangled names. How does this sound?
We've known about `calli` for years now but it involves a large amount of work since, as you said, the instruction isn't available in C#. The issue is further complicated by the generated code still relatively often having errors to manually correct which precludes the automatic compilation necessary for rewriting of IL. Therefore this would be a significant challenge.
I've wondered about `ICustomMarshaller` myself. We don't use it anywhere which isn't deliberate but rather just a habit. It might as well be a superior solution, including in performance. @tritao what do you think? Could we perhaps complement or replace type maps with custom marshallers?